pkg/stanza/fileconsumer/file_threadpool.go

```diff
@@ -0,0 +1,198 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package fileconsumer // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer"
+
+import (
+	"context"
+	"os"
+	"sync"
+
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/fingerprint"
+)
+
+type readerWrapper struct {
+	reader *reader
+	fp     *fingerprint.Fingerprint
+}
+
+func (m *Manager) kickoffThreads(ctx context.Context) {
+	m.readerChan = make(chan readerWrapper, m.maxBatchFiles*2)
+	for i := 0; i < m.maxBatchFiles; i++ {
+		m.workerWg.Add(1)
+		go m.worker(ctx)
+	}
+}
+
+func (m *Manager) shutdownThreads() {
+	if m.readerChan != nil {
+		close(m.readerChan)
+	}
+	m.workerWg.Wait()
+	// save off any files left
+	// As we already cancelled our current context, create a new one to save any left offsets
+	// This is only applicable for `filelog.useThreadPool` featuregate
+	ctx, cancel := context.WithCancel(context.Background())
+	m.syncLastPollFilesConcurrent(ctx)
+	cancel()
+}
+
+// poll checks all the watched paths for new entries
+func (m *Manager) pollConcurrent(ctx context.Context) {
+	// Increment the generation on all known readers
+	// This is done here because the next generation is about to start
+	m.knownFilesLock.Lock()
```

Comment:
@djaglowski I _guess_ we can remove locking from this spot? I wasn't sure about it that's why I pushed it.
We update each reader's generation, we don't append/delete the array in any form.